### PR TITLE
Initial Solution Pack Merge

### PR DESCRIPTION
**Pull request by fortijames**
This is the first release of the solution pack for review. Going forward I'll continue to fork the ( https://github.com/fortinet-fortisoar/solution-pack-fortimanager-ztp-framework ) repo and push changes. I also intend to leverage ( https://github.com/fortinet-fortisoar/solution-pack-fortimanager-ztp-framework/issues ) for bugs and features.

### CODE REVIEW

File: connectors/data.json [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        "operation_roles": [],
+        "configurations": [
+            {
+                "id": 31,
+                "config_id": "6452db36-dc11-44a1-9a7f-231a9b569ef5",
+                "status": 1,
+                "name": "Default",
+                "default": true,
+                "config": {
+                    "allow_imports": true
```
Comment (cs-rohit-khune) [inline connectors/data.json]: @fortijames Configuring the code snippet connector exclusively with the libraries used in the content pack is advisable. Allowing all library imports is not recommended.

File: README.md [hunk 2]
```diff
@@ -1 +1,37 @@
-# solution-template-repository
\ No newline at end of file
+# Release Information
+ - **Version**: 1.0.3
```
Comment (cs-rohit-khune) [inline README.md]: @fortijames Please make sure to set the version to 1.0.0 and commit your solution pack code to the release/1.0.0 branch instead of the develop branch.

File: connectors/data.json [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        ],
+        "dataImports": [],
+        "install_mode": "rpm",
+        "ingestionPlaybooks": {
+            "collections": [],
+            "exported_tags": []
+        }
+    },
+    {
+        "name": "fortinet-fortimanager_dev",
```
Comment (cs-rohit-khune) [inline connectors/data.json:14]: @fortijames Please ensure that the released version of the connector is added. Instead of "fortinet-fortimanager_dev", please add "fortinet-fortimanager".

File: connectors/data.json [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        "label": "Fortinet FortiManager Custom",
+        "category": "Centralized Security Management",
+        "description": "The Fortinet FortiManager provides easy centralized configuration, policy-based provisioning, update management and end-to-end network monitoring for your Fortinet installed environment.",
+        "publisher": null,
+        "version": "3.0.0",
```
Comment (cs-rohit-khune) [inline connectors/data.json:19]: - do not mention the version in JSON, please remove this key

File: README.md [hunk 5]
```diff
@@ -1 +1,37 @@
...
+ - **Certified**: unknown
```
Comment (cs-rohit-khune) [inline README.md]: Mark it as "No" instead of "unknown".

File: README.md (hunk 2 above)
Comment (cs-rohit-khune) [reply README.md]: @fortijames Please create documentation using the standard format provided in  https://github.com/fortinet-fortisoar/how-tos/tree/main/documenting

File: dashboards/FortiManager Device Metafield Templates.json [hunk 6]
```diff
@@ -0,0 +1,105 @@
+{
+    "@context": "\/api\/3\/contexts\/Dashboard",
+    "@type": "Dashboard",
+    "uuid": "79ea6ac2-f1a5-4287-a17e-988c15f8661c",
+    "displayName": "FortiManager Device Metafield Templates",
+    "type": "rows",
+    "templateType": "dashboard",
+    "roles": [],
```
Comment (cs-rohit-khune) [inline dashboards/FortiManager Device Metafield Templates.json:8]: @fortijames Please ensure that a role is assigned to the dashboard.

File: dashboards/FortiManager Metadata Assignments.json [hunk 7]
```diff
@@ -0,0 +1,15 @@
+{
+    "@context": "\/api\/3\/contexts\/Dashboard",
+    "@type": "Dashboard",
+    "uuid": "252d33ea-0a20-4421-b87a-30e2dfe172b4",
+    "displayName": "FortiManager Metadata Assignments",
+    "type": "rows",
+    "templateType": "dashboard",
+    "roles": [],
```
Comment (cs-rohit-khune) [inline dashboards/FortiManager Metadata Assignments.json:8]: Please ensure that a role is assigned to the dashboard. @fortijames

File: dashboards/FortiManager Product Details.json [hunk 8]
```diff
@@ -0,0 +1,234 @@
+{
+    "@context": "\/api\/3\/contexts\/Dashboard",
+    "@type": "Dashboard",
+    "uuid": "d9772022-ba8b-4351-a707-7b1df8d4eeaf",
+    "displayName": "FortiManager Product Details",
+    "type": "rows",
+    "templateType": "dashboard",
+    "roles": [],
```
Comment (cs-rohit-khune) [inline dashboards/FortiManager Product Details.json]: @fortijames Please ensure that a role is assigned to the dashboard.

File: dashboards/FortiManager ZTP Profile and Phase Summary.json [hunk 9]
```diff
@@ -0,0 +1,385 @@
+{
+    "@context": "\/api\/3\/contexts\/Dashboard",
+    "@type": "Dashboard",
+    "uuid": "88827bf8-2014-4881-8b28-8ece441bdddd",
+    "displayName": "FortiManager ZTP Profile and Phase Summary",
+    "type": "rows",
+    "templateType": "dashboard",
+    "roles": [],
```
Comment (cs-rohit-khune) [inline dashboards/FortiManager ZTP Profile and Phase Summary.json]: @fortijames Please ensure that a role is assigned to the dashboard.

File: dashboards/FortiManager ZTPF Docs - Metafield Sources.json [hunk 10]
```diff
@@ -0,0 +1,79 @@
+{
+    "@context": "\/api\/3\/contexts\/Dashboard",
+    "@type": "Dashboard",
+    "uuid": "9545ae4e-0552-43e4-a399-1082dabe8a98",
+    "displayName": "FortiManager ZTPF Docs - Metafield Sources",
+    "type": "rows",
+    "templateType": "dashboard",
+    "roles": [],
```
Comment (cs-rohit-khune) [inline dashboards/FortiManager ZTPF Docs - Metafield Sources.json]: Please ensure that a role is assigned to the dashboard.

File: dashboards/FortiManager ZTPF Docs - ZTP Profiles.json [hunk 11]
```diff
@@ -0,0 +1,79 @@
+{
+    "@context": "\/api\/3\/contexts\/Dashboard",
+    "@type": "Dashboard",
+    "uuid": "d61bc75d-ad4b-494f-93ab-62ddc78ed466",
+    "displayName": "FortiManager ZTPF Docs - ZTP Profiles",
+    "type": "rows",
+    "templateType": "dashboard",
+    "roles": [],
```
Comment (cs-rohit-khune) [inline dashboards/FortiManager ZTPF Docs - ZTP Profiles.json]: @fortijames Please ensure that a role is assigned to the dashboard.

File: info.json [hunk 12]
```diff
@@ -0,0 +1,223 @@
+{
+    "name": "fortiManagerZTPFlow",
+    "version": "1.0.5",
```
Comment (cs-rohit-khune) [inline info.json]: @fortijames The version should be 1.0.0

File: info.json [hunk 13]
```diff
@@ -0,0 +1,223 @@
...
+    "fsrMinCompatibility": "7.4.0",
+    "type": "solutionpack",
+    "local": true,
+    "label": "FortiManager ZTP Flow",
+    "dependencies": [],
```
Comment (cs-rohit-khune) [inline info.json:8]: @fortijames Make sure to add any solution pack dependency such as "SOAR Framework"

File: info.json [hunk 14]
```diff
@@ -0,0 +1,223 @@
...
+    "publisher": "Fortinet",
```
Comment (cs-rohit-khune) [inline info.json:10]: @fortijames publisher should be "Fortinet CSE"

File: info.json [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+    "description": "A solution framework to manage one or more FortiManagers and attached Fortinet Devices. Build a ZTP Flow that will perform onboarding, lifecycle, and/or offboarding objectives for devices. Create a single pane of glass dashboard for multiple FortiManagers and attached devices. Create reports for devices using jinja to simplify data collection. Interact with external users and/or systems to collect device metadata required for various tasks.",
+    "help": "https://github.com/fortinet-fortisoar/solution-pack-fortimanager-ztp-framework",
```
Comment (cs-rohit-khune) [inline info.json:12]: Please provide a link to README.md on the release/1.0.0 branch. i.e

https://github.com/fortinet-fortisoar/solution-pack-fortimanager-ztp-framework/blob/release/1.0.0/README.md

File: info.json [hunk 16]
```diff
@@ -0,0 +1,223 @@
...
+                "name": "FortiManager(FMG)-Synchronizing"
+            },
+            {
+                "name": "FortiManager(FMG)-Triggers"
+            }
+        ],
+        "globalVariables": [
+            {
+                "id": 7,
+                "name": "Server_fqhn",
```
Comment (cs-rohit-khune) [inline info.json:165]: @fortijames Please remove the global variable 'Server_fqdn' as its value may vary across different FSR instances.

File: info.json [hunk 17]
```diff
@@ -0,0 +1,223 @@
...
+                "name": "FortiManager ZTP Profile and Phase Summary"
+            }
+        ],
+        "connectors": [
+            {
+                "name": "Code Snippet",
+                "apiName": "code-snippet"
+            },
+            {
+                "name": "Fortinet FortiManager Custom",
```
Comment (cs-rohit-khune) [inline info.json:219]: @fortijames please mention released released connector here

File: roles/FortiManager-Admin-c57340f1-c387-474b-9c24-effc9e3c41aa.json [hunk 18]
```diff
@@ -0,0 +1,585 @@
+{
+    "@context": "\/api\/3\/contexts\/Role",
+    "@id": "\/api\/3\/roles\/c57340f1-c387-474b-9c24-effc9e3c41aa",
+    "@type": "Role",
+    "description": "The FortiManager Admin should be able to edit only certain fields as the synchronization process owns many fields.",
+    "name": "FortiManager-Admin",
+    "modulePermissions": [
```
Comment (cs-rohit-khune) [inline roles/FortiManager-Admin-c57340f1-c387-474b-9c24-effc9e3c41aa.json:7]: @fortijames Please include only the custom modules that are added with this solution pack in the role. Exclude Out-of-the-Box (OOB) modules such as agents, alerts, and so on.

File: roles/FortiManager-Playbook-c064238d-8049-4c5b-b3e8-e3a1c41178fe.json [hunk 19]
```diff
@@ -0,0 +1,531 @@
+{
+    "@context": "\/api\/3\/contexts\/Role",
```
Comment (cs-rohit-khune) [inline roles/FortiManager-Playbook-c064238d-8049-4c5b-b3e8-e3a1c41178fe.json:2]: @fortijames Please include only the custom modules that are added with this solution pack in the role. Exclude Out-of-the-Box (OOB) modules such as agents, alerts, and so on.

**Review by cs-rohit-khune — CHANGES_REQUESTED** [review]

File: playbooks/FortiManager(FMG)-Triggers/On ZTP Profile Assignment in Device Record.json [hunk 20]
```diff
@@ -0,0 +1,557 @@
+{
+    "@type": "Workflow",
+    "triggerLimit": null,
+    "name": "On ZTP Profile Assignment in Device Record",
```
Comment (cs-rohit-khune) [inline playbooks/FortiManager(FMG)-Triggers/On ZTP Profile Assignment in Device Record.json:4]: @fortijames Trigger condition is missing in trigger step 

image

**Review by cs-rohit-khune — CHANGES_REQUESTED** [review]

**Comment by cs-rohit-khune** [comment]
@fortijames Although already mentioned in the pre-requisite instructions, it is recommended to add a note in the documentation that instructs users to assign the 'FortiManager-Playbook' and 'FortiManager-Admin' roles.

File: connectors/data.json (hunk 3 above)
Comment (fortijames) [reply connectors/data.json:14]: > @fortijames Please ensure that the released version of the connector is added. Instead of "fortinet-fortimanager_dev", please add "fortinet-fortimanager".

This one is pending a connector that works with the solution pack. You just created the one we will be using. 
https://github.com/fortinet-fortisoar/connector-fortinet-fortimanager-json-rpc

My thinking here was to get things rolling so we have something started and then push updates to fix the dev connector. Anyone that uses this solution pack as a pre-release would need to download the dev connector and I can put that in the setup docs. I mentioned this in email as a step required before publishing to the Content Hub. 

In short. I expect this to be changed in the next PR and did not expect it to stop the merge of the rest of the SP.

File: README.md (hunk 5 above)
Comment (fortijames) [reply README.md]: > Mark it as "No" instead of "unknown".

Ok, how do we get it Certified? Is there a doc on this?

**Comment by fortijames** [comment]
What's the process for addressing comments in the PR? 
 - Is there a way to update my PR code with my latest changes? I did not see that option. 
 - Close this PR and create a new? 
 - Create a new PR with additions and have you merge/play them in order?

File: README.md (hunk 2 above)
Comment (fortijames) [reply README.md]: > @fortijames Please create documentation using the standard format provided in https://github.com/fortinet-fortisoar/how-tos/tree/main/documenting

Is there a tool for creating `contents.md`? That would be handy. :-)

File: connectors/data.json (hunk 1 above)
Comment (fortijames) [reply connectors/data.json]: > @fortijames Configuring the code snippet connector exclusively with the libraries used in the content pack is advisable. Allowing all library imports is not recommended.

I just moved this to a setup requirement for now. We can cleanup later.

File: connectors/data.json (hunk 4 above)
Comment (fortijames) [reply connectors/data.json:19]: > * do not mention the version in JSON, please remove this key

This connector is in the process of being replaced.

File: info.json (hunk 15 above)
Comment (fortijames) [reply info.json:12]: > Please provide a link to README.md on the release/1.0.0 branch. i.e
> 
> https://github.com/fortinet-fortisoar/solution-pack-fortimanager-ztp-framework/blob/release/1.0.0/README.md

Since I don't know the best path here. I will cancel my PR and redo as a 1.0.0 branch request.

File: info.json (hunk 16 above)
Comment (fortijames) [reply info.json:165]: > @fortijames Please remove the global variable 'Server_fqdn' as its value may vary across different FSR instances.

I don't know how this is being exported. I have Global Vars unchecked.

File: info.json (hunk 17 above)
Comment (fortijames) [reply info.json:219]: > @fortijames please mention released released connector here

I assume you mean the `Fortinet FortiManager Custom`. This connector is being replaced.

**Comment by fortijames** [comment]
I'm closing and will try again. Please let me know the proper flow for making changes within a PR.

File: README.md (hunk 2 above)
Comment (cs-rohit-khune) [reply README.md]: No, there is no tool available for this
